apps/prairielearn/src/ee/pages/lti13CourseNavigation/lti13CourseNavigation.ts

```diff
@@ -1,10 +1,189 @@
+          deployment_id: LTI.deployment_id,
+          context_id: LTI.context?.id,
+          context_label: LTI.context?.label,
+          context_title: LTI.context?.title,
+        });
+
+        // TODO: Set course/instance staff permissions for LMS course staff here?
+      }
+
+      // LTI claims are not used after this page so remove them from the session
+      LTI.remove();
+
+      // Redirect to linked course instance
+      res.redirect(
+        `/pl/course_instance/${lci.course_instance_id}/${role_instructor ? 'instructor/' : ''}`,
+      );
+      return;
+    }
+
+    if (!role_instructor) {
+      // Students get a "come back later" message
+
+      /*
+       * TODO: to not break things for current testing courses,
+       *       fall back to the previous PR behavior of simply
+       *       redirecting to /pl
+       */
+      res.redirect('/pl/');
+      return;
+
+      /*
+      res.send(
+        Lti13CourseNavigationNotReady({
+          resLocals: res.locals,
+          courseName,
+        }),
+      );
+      return;
+      */
+    }
+
+    // Instructor so lookup their existing information in PL
+    const courses = await selectCoursesWithEditAccess({
+      user_id: res.locals.authn_user.user_id,
+      is_administrator: res.locals.authn_is_administrator,
+    });
+
+    let course_instances: CourseInstance[] = [];
+
+    for (const course of courses) {
+      // Only course owners can link
+      if (!course.permissions_course.has_course_permission_own) {
+        continue;
+      }
+
+      const loopCI = await selectCourseInstancesWithStaffAccess({
+        course_id: course.id,
+        user_id: res.locals.authn_user.user_id,
+        authn_user_id: res.locals.authn_user.user_id,
+        is_administrator: res.locals.authn_is_administrator,
+        authn_is_administrator: res.locals.authn_is_administrator,
+      });
+
+      course_instances = [...course_instances, ...loopCI];
+    }
+
+    res.send(
+      Lti13CourseNavigationInstructor({
+        resLocals: res.locals,
+        courseName,
+        courses,
+        course_instances,
+      }),
+    );
+  }),
+);
+
+router.post(
+  '/',
+  asyncHandler(async (req, res) => {
+    const unsafe_course_instance_id = req.body.ci_id;
+    const unsafe_lti13_instance_id = req.params.lti13_instance_id;
+
+    const LTI = new Lti13Claim(req);
+    const authn_lti13_instance_id = req.session.authn_lti13_instance_id;
+
+    // Validate user login matches this lti13_instance
+    if (unsafe_lti13_instance_id !== authn_lti13_instance_id) {
+      throw error.make(403, 'Permission denied');
+    }
+
+    try {
+      // Map lti13_instance through institution to course instance or fail
+      await queryOneRowAsync(sql.check_lti_ci, {
+        course_instance_id: req.body.ci_id,
+        lti13_instance_id: req.params.lti13_instance_id,
+      });
+    } catch {
+      throw error.make(403, 'Permission denied');
+    }
```

Comment:
I'd rather avoid exceptions as control flow, as this makes us impossible to differentiate between "row not found" and "something else went wrong" (query is broken, database is dead, etc). Instead, use `queryOptionalRow`, then you can throw your error if the result is `null`.